File: Java-Core/src/main/java/presto/parser/AbstractParser.java
package presto.parser;

import java.util.List;

import org.antlr.v4.runtime.BufferedTokenStream;
import org.antlr.v4.runtime.Parser;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.TokenStream;

public abstract class AbstractParser extends Parser {

	public AbstractParser(TokenStream input) { 
		super(input);
	}
	
	public boolean isText(Token token, String text) {
		return text.equals(token.getText());
	}
	
	public boolean was(int type) {
		return lastHiddenTokenType()==type;
	}

	public boolean wasNot(int type) {
		return lastHiddenTokenType()!=type;
	}
	
	public boolean willBe(int type) {
		return getTokenStream().LA(1)==type;
	}
	
	public boolean willNotBe(int type) {
		return getTokenStream().LA(1)!=type;
	}
	
	public int equalToken() {
		throw new RuntimeException("You must override equalToken() !");
	};

	public boolean willBeAOrAn() {
		return willBeText("a") || willBeText("an");
	}
	
	public boolean willBeText(String text) {
		return text.equals(getTokenStream().LT(1).getText());
	}

	public int nextHiddenTokenType() {
		BufferedTokenStream bts = (BufferedTokenStream)getTokenStream();
		List<Token> hidden = bts.getHiddenTokensToRight(bts.index()-1);
		if(hidden==null || hidden.size()==0)
			return 0;
		else
			return hidden.get(0).getType();
	}
	
	public int lastHiddenTokenType() {
		BufferedTokenStream bts = (BufferedTokenStream)getTokenStream();
		List<Token> hidden = bts.getHiddenTokensToLeft(bts.index());
		if(hidden==null || hidden.size()==0)
			return 0;
		else
			return hidden.get(hidden.size()-1).getType();
	}
	
}

File: Java-CodeServer/src/main/resources/js/prompto/parser/AbstractParser.js
var Parser = require("antlr4/index").Parser;

function AbstractParser(input) {
	Parser.call(this, input);
	return this;
}

AbstractParser.prototype = Object.create(Parser.prototype);
AbstractParser.prototype.constructor = AbstractParser;

AbstractParser.prototype.isText = function(token, text) {
	return text === token.text;
};

AbstractParser.prototype.was = function(type) {
	return this.lastHiddenTokenType()===type;
};

AbstractParser.prototype.wasNot = function(type) {
	return this.lastHiddenTokenType()!==type;
};

AbstractParser.prototype.willBe = function(type) {
	return this.getTokenStream().LA(1)===type;
};

AbstractParser.prototype.willNotBe = function(type) {
	return this.getTokenStream().LA(1)!==type;
};

AbstractParser.prototype.nextHiddenTokenType = function() {
	var bts = this.getTokenStream();
	var hidden = bts.getHiddenTokensToRight(bts.index-1);
	if(hidden===null || hidden.length===0) {
		return 0;
	} else {
		return hidden[0].type;
	}
};

AbstractParser.prototype.willBeAOrAn = function() {
    return this.willBeText("a") || this.willBeText("an");
};

AbstractParser.prototype.willBeText = function(text) {
    return text===this.getTokenStream().LT(1).text;
};

AbstractParser.prototype.lastHiddenTokenType = function() {
	var bts = this.getTokenStream();
	var hidden = bts.getHiddenTokensToLeft(bts.index);
	if(hidden===null || hidden.length===0) {
		return 0;
	} else {
		return hidden[hidden.length-1].type;
	}
};

exports.AbstractParser = AbstractParser;